src/structured-data/handler.js

```diff
@@ -0,0 +1,132 @@
+ * @param {Array} pages - An array of page URLs to be audited.
+ *
+ * @returns {Promise<Array<Object>>} - A promise that resolves to an array of objects,
+ * each containing the inspection URL, filtered index status result, and filtered rich results.
+ * If an error occurs during the inspection of a URL, the object will include an error message.
+ *
+ * @throws {Error} - Throws an error if the audit process fails.
+ */
+export async function processStructuredData(baseURL, context, pages) {
+  const { log } = context;
+
+  let google;
+  try {
+    google = await GoogleClient.createFrom(context, baseURL);
+  } catch (error) {
+    log.error(`Failed to create Google client. Site was probably not onboarded to GSC yet. Error: ${error.message}`);
+    throw new Error(`Failed to create Google client. Site was probably not onboarded to GSC yet. Error: ${error.message}`);
+  }
+
+  const urlInspectionResult = pages.map(async (page) => {
+    try {
+      const { inspectionResult } = await google.urlInspect(page);
+      log.info(`Successfully inspected URL: ${page}`);
+
+      const filteredIndexStatusResult = {
+        verdict: inspectionResult?.indexStatusResult?.verdict,
+        lastCrawlTime: inspectionResult?.indexStatusResult?.lastCrawlTime,
+      };
+
+      const detectedItemTypes = [];
+      const filteredRichResults = inspectionResult?.richResultsResult?.detectedItems?.map(
+        (item) => {
+          detectedItemTypes.push(item?.richResultType);
+          const filteredItems = item?.items?.filter(
+            (issueItem) => issueItem?.issues?.some(
+              (issue) => issue?.severity === 'ERROR',
+            ),
+          )?.map((issueItem) => ({
+            name: issueItem?.name,
+            issues: issueItem?.issues?.filter((issue) => issue?.severity === 'ERROR'),
+          }));
+
+          return {
+            richResultType: item?.richResultType,
+            items: filteredItems,
+          };
+        },
+      )?.filter((item) => item.items.length > 0) ?? [];
+
+      if (filteredRichResults.length > 0) {
+        filteredRichResults.verdict = inspectionResult?.richResultsResult?.verdict;
+        log.info(`Found ${filteredRichResults.length} rich results issues for URL: ${page}`);
+      } else {
+        log.info(`No rich results issues found for URL: ${page}`);
+      }
+      return {
+        inspectionUrl: page,
+        indexStatusResult: filteredIndexStatusResult,
+        richResults: inspectionResult?.richResultsResult
+          ? {
+            verdict: inspectionResult.richResultsResult.verdict,
+            detectedItemTypes,
+            detectedIssues: filteredRichResults,
+          }
+          : {},
+      };
+    } catch (error) {
+      log.error(`Failed to inspect URL: ${page}. Error: ${error.message}`);
+      return {
+        inspectionUrl: page,
+        error: error.message,
+      };
+    }
+  });
+
+  return Promise.all(urlInspectionResult);
```

Comment:
I would use`Promise.allSettled`instead, to ensure that all the promises are handled individually. with Promise.all in case one promise is rejected  it rejects the entire operation

Reply:
good point, thanks!